modules/redis/src/main/scala/zio/redis/internal/SingleNodeSubscriptionExecutor.scala

```diff
@@ -0,0 +1,221 @@
+      case Subscription.Subscribe =>
+        subscribe(extractChannelKeys(command.args), command)
+      case Subscription.PSubscribe =>
+        subscribe(extractPatternKeys(command.args), command)
+      case Subscription.Unsubscribe =>
+        unsubscribe(command).as(ZStream.empty)
+      case Subscription.PUnsubscribe =>
+        unsubscribe(command).as(ZStream.empty)
+      case other => ZIO.fail(RedisError.InvalidPubSubCommand(other))
+    }
+
+    for {
+      commandName <- ZStream.fromZIO(
+                       ZIO
+                         .fromOption(command.args.collectFirst { case RespCommandArgument.CommandName(name) =>
+                           name
+                         })
+                         .orElseFail(RedisError.CommandNameNotFound(command))
+                     )
+      pushMessage <- ZStream.fromZIO(getStream(commandName)).flatten
+    } yield pushMessage
+  }
+
+  private def subscribe(
+    keys: Chunk[SubscriptionKey],
+    command: RespCommand
+  ): IO[RedisError, Stream[RedisError, PushMessage]] = {
+    def getHub(key: SubscriptionKey): IO[RedisError, Hub[Take[RedisError, PushMessage]]] =
+      subscriptionMap
+        .get(key)
+        .flatMap(ZIO.fromOption(_))
+        .orElseFail(RedisError.SubscriptionStreamAlreadyClosed(key))
+
+    def getStream(promise: Promise[RedisError, PushMessage]) =
+      for {
+        subscribed <- ZStream.fromZIO(promise.await)
+        head        = ZStream(subscribed)
+        tail        = ZStream.fromZIO(getHub(subscribed.key)).flatMap(ZStream.fromHub(_)).flattenTake
+        message    <- head ++ tail
+      } yield message
+
+    def makeHub(key: SubscriptionKey): UIO[Unit] =
+      Hub
+        .unbounded[Take[RedisError, PushMessage]]
+        .tap(subscriptionMap.putIfAbsent(key, _))
+        .unit
+
+    for {
+      _        <- ZIO.foreachDiscard(keys)(makeHub)
+      promises <- Promise.make[RedisError, PushMessage].replicateZIO(keys.size).map(Chunk.fromIterable)
+      _        <- requests.offer(Request.Subscribe(command.args.map(_.value), promises))
+      streams   = promises.map(getStream)
+    } yield streams.fold(ZStream.empty)(_ merge _)
+  }
+
+  private def unsubscribe(command: RespCommand): IO[RedisError, Unit] =
+    for {
+      promise <- Promise.make[RedisError, Unit]
+      _       <- requests.offer(Request.Unsubscribe(command.args.map(_.value), promise))
+      _       <- promise.await
+    } yield ()
+
+  private def extractChannelKeys(command: Chunk[RespCommandArgument]): Chunk[SubscriptionKey.Channel] =
+    command.collect { case RespCommandArgument.Key(key) =>
+      key.asString
+    }.map(SubscriptionKey.Channel.apply)
+
+  private def extractPatternKeys(command: Chunk[RespCommandArgument]): Chunk[SubscriptionKey.Pattern] =
+    command.collect { case RespCommandArgument.Key(key) =>
+      key.asString
+    }.map(SubscriptionKey.Pattern.apply)
+
+  def send: IO[RedisError.IOError, Unit] =
+    requests.takeAll.flatMap { reqs =>
```

Comment:
Use `takeBetween(1, RequestQueueSize)` just like in the existing `send` [implementation](https://github.com/zio/zio-redis/blob/master/modules/redis/src/main/scala/zio/redis/internal/SingleNodeExecutor.scala#L49).